File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    List<String> res = new ArrayList<>();
    public List<String> generateParenthesis(int n) {
        
        StringBuilder sb = new StringBuilder();
        int open = 0;
        int close = 0;
        backtrack(sb,res,n,open,close);
        return res;

    }
    public void backtrack(StringBuilder sb, List<String> res, int n, int open, int close){
        if(open == n && open == close){
            res.add(sb.toString());
            return;
        }
        if(open < n){
            sb.append('(');
            backtrack(sb,res,n,open+1,close);
            sb.deleteCharAt(sb.length()-1);//backtrakcing
        }
        if(open > close){
            sb.append(')');
            backtrack(sb, res, n, open, close+1);
            sb.deleteCharAt(sb.length()-1);
        }
    }
}

File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack=[]
        res=[]
        def backtrack(closeN,openN):
            if closeN==openN==n:#base case
                res.append("".join(stack))
                return
            if openN<n:#open bracket whenever open < n
                stack.append('(')
                backtrack(closeN,openN+1)
                stack.pop()
            if closeN<openN: # close braket only when close less than open
                stack.append(')')
                backtrack(closeN+1,openN)
                stack.pop()
            
        backtrack(0,0)
        return res